### DISCUSSION TITLE
Backport Next 15 features that don't need React 19 RC

### DISCUSSION THREAD

**Discussion by flexdinesh** [vercel/next.js#73405] [App Router]
Next 15 has a few features that are not related to React 19 but are released in Next 15, which requires React 19.
Some of these features include:
- [Support for TypeScript in next config](https://nextjs.org/blog/next-15#support-for-nextconfigts)
- [Custom cache headers and other self-hosting improvements](https://nextjs.org/blog/next-15#improvements-for-self-hosting)
- [Performance improvements](https://nextjs.org/blog/next-15#development-and-build-improvements)
- [Turbopack Dev](https://nextjs.org/blog/next-15#turbopack-dev)

Points to help with the discussion:
- React 19 hasn't been released as a stable major yet. 
- [React 19 RC has plenty of breaking changes](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#breaking-changes). 
- React open-source ecosystem packages are not yet ready to support React 19 because React 19 is still in RC

Upgrading to React 19 is not an easy task, especially for people working in big monorepos with many ecosystem packages. The ecosystem is not ready for React 19 yet because React 19 is still in RC. For example, one of the work projects is set up with a monorepo with a shared design system that's built with primitives using open-source packages. We cannot upgrade the design system to use React 19 yet because 1) it's still in RC, 2) the packages we use don't support React 19 yet. Because of that, the next.js workspace cannot upgrade to Next 15 because React 19 is a requirement for Next 15. That's okay because some Next 15 features are dependent on React 19.

But why are all the new features released with Next 15? For example, we're self-hosting, and we've been waiting for a long time to use 1) typescript in next config (we have scripts that run during build to set up the environment) 2) customise cache control headers in the page response. But since we can't upgrade to React 19, we can't make use of the new features in Next 15 that are not related to React 19 either. Practically speaking, we are probably looking at at least a year before the ecosystem catches up to React 19 and our design system is upgraded to React 19.

So here are my questions:
1. Is there a chance that some of the Next 15 features will be backported to Next 14? 
2. For those that are in a monorepo with a next.js workspace and shared packages that use React 18, what's your upgrade strategy for Next 15 look like?

**Comment by leerob** [comment]
>  the packages we use don't support React 19 yet

Could you share the specific packages please?

**Reply by flexdinesh** [reply]
Sorry, I don't have the specifics yet, as I haven't dug deep into the errors. Some of the errors we face are mostly from internal packages that need to be upgraded to support React 19. But that's something we can work out with other teams to fix. But if these errors are from upstream packages (for example, framer-motion) we can't really fix that now.

This is one such error that happens when we load the page. We haven't debugged it yet, as we're not sure if it's from an internal package or from an upstream dependency.

> TypeError: Cannot read properties of null (reading 'useRef')

We're also looking at lots of peer dependency mismatches. We haven't noticed any build failures when we ignore the peer dependency mismatch. Some packages, for eg. [radix-ui](https://github.com/radix-ui/primitives/pull/2811) and [sentry](https://github.com/getsentry/sentry-javascript/blob/12d5da6328a850f0398c288dadc3f71282c5dde6/packages/react/package.json#L47) already support React 19 RC. But not all packages do because the package maintainers will need to test and update their packages to include React 19 RC as a peer dependency, which not many folks in the ecosystem haven't gotten around to yet because React 19 stable major hasn't been released yet. Some examples [framer-motion](https://github.com/motiondivision/motion/issues/2668#issue-2299682053), [paypal-js](https://github.com/paypal/paypal-js/issues/574), etc. For consumers, if we ignore the warning and then something breaks, we can't even file an issue because we're on a peer version that's not supported by the upstream package.

```
for many react ecosystem packages
✕ unmet peer react@"^16.8.0 || ^17.0.0-rc.1 || ^18.0.0": found 19.0.0-rc.1
```

Given that all the internal packages, design system, etc will need to be upgraded to React 19 at one point. We'd rather we do that when the ecosystem is ready and not get ahead of ourselves by ignoring the peer warnings. There are lots of goodies in Next 15, especially the improvements for self-hosting. Some of these features (from my understanding of the feature) don't require React 19. It's a bummer that we can't use these useful features yet, even though they're not related to React 19.

**Reply by leerob** [reply]
Did you already go through the upgrade guide? There are some changes around how refs are handled. 

https://react.dev/blog/2024/04/25/react-19-upgrade-guide

framer-motion works with React 19 (we are using it on a number of projects). I haven't tested the paypal library yet. Any others?
